In [ ]:
# Import the modules
import numpy as np
import pandas as pd
from pathlib import Path
from sklearn.metrics import balanced_accuracy_score, confusion_matrix, classification_report

---

## Split the Data into Training and Testing Sets

### Step 1: Read the `lending_data.csv` data from the `Resources` folder into a Pandas DataFrame.

In [ ]:
# Read the CSV file into a Pandas DataFrame
# YOUR CODE HERE!

credit_risk_df = pd.read_csv("lending_data.csv")

# Review the DataFrame
# YOUR CODE HERE!
credit_risk_df.head()

loan_size  interest_rate  borrower_income  debt_to_income  num_of_accounts  \
0    10700.0          7.672            52800        0.431818                5   
1     8400.0          6.692            43600        0.311927                3   
2     9000.0          6.963            46100        0.349241                3   
3    10700.0          7.664            52700        0.430740                5   
4    10800.0          7.698            53000        0.433962                5   

   derogatory_marks  total_debt  loan_status  
0                 1       22800            0  
1                 0       13600            0  
2                 0       16100            0  
3                 1       22700            0  
4                 1       23000            0

### Step 2: Create the labels set (`y`)  from the “loan_status” column, and then create the features (`X`) DataFrame from the remaining columns.

In [ ]:
# Separate the data into labels and features
# Separate the y variable, the labels
# YOUR CODE HERE!]
y = credit_risk_df["loan_status"]

# Separate the X variable, the features
# YOUR CODE HERE!
X = credit_risk_df.drop(columns="loan_status")

In [ ]:
# Review the y variable Series
# YOUR CODE HERE!
y[:5]

0    0
1    0
2    0
3    0
4    0
Name: loan_status, dtype: int64

In [ ]:
# Review the X variable DataFrame
# YOUR CODE HERE!
new_credit_risk_df = pd.DataFrame(X)

new_credit_risk_df

loan_size  interest_rate  borrower_income  debt_to_income  \
0        10700.0          7.672            52800        0.431818   
1         8400.0          6.692            43600        0.311927   
2         9000.0          6.963            46100        0.349241   
3        10700.0          7.664            52700        0.430740   
4        10800.0          7.698            53000        0.433962   
...          ...            ...              ...             ...   
77531    19100.0         11.261            86600        0.653580   
77532    17700.0         10.662            80900        0.629172   
77533    17600.0         10.595            80300        0.626401   
77534    16300.0         10.068            75300        0.601594   
77535    15600.0          9.742            72300        0.585062   

       num_of_accounts  derogatory_marks  total_debt  
0                    5                 1       22800  
1                    3                 0       13600  
2                    3                 0       16100  
3                    5                 1       22700  
4                    5                 1       23000  
...                ...               ...         ...  
77531               12                 2       56600  
77532               11                 2       50900  
77533               11                 2       50300  
77534               10                 2       45300  
77535                9                 2       42300  

[77536 rows x 7 columns]

### Step 3: Check the balance of the labels variable (`y`) by using the `value_counts` function.

In [ ]:
# Check the balance of our target values
# YOUR CODE HERE!
target_values = pd.Series(y).value_counts()
target_values

loan_status
0    75036
1     2500
Name: count, dtype: int64

### Step 4: Split the data into training and testing datasets by using `train_test_split`.

In [ ]:
# Import the train_test_learn module
from sklearn.model_selection import train_test_split

# Split the data using train_test_split
# Assign a random_state of 1 to the function
# YOUR CODE HERE!
X_train, X_test, y_train, y_test = train_test_split(X, y, random_state=1)

---

## Create a Logistic Regression Model with the Original Data

###  Step 1: Fit a logistic regression model by using the training data (`X_train` and `y_train`).

In [ ]:
# Import the LogisticRegression module from SKLearn
from sklearn.linear_model import LogisticRegression

# Instantiate the Logistic Regression model
# Assign a random_state parameter of 1 to the model
# YOUR CODE HERE!
model = LogisticRegression(random_state = 1)

# Fit the model using training data
# YOUR CODE HERE!
model.fit(X_train,y_train)

model

LogisticRegression(random_state=1)

### Step 2: Save the predictions on the testing data labels by using the testing feature data (`X_test`) and the fitted model.

In [ ]:
# Make a prediction using the testing data
# YOUR CODE HERE!
predictions = model.predict(X_test)

### Step 3: Evaluate the model’s performance by doing the following:

* Calculate the accuracy score of the model.

* Generate a confusion matrix.

* Print the classification report.

In [ ]:
# Print the balanced_accuracy score of the model
# YOUR CODE HERE!
accuracy_score = balanced_accuracy_score(y_test, predictions)
print(f"Accuracy Score: {accuracy_score * 100:.2f}%")

Accuracy Score: 96.80%


In [ ]:
# Generate a confusion matrix for the model
# YOUR CODE HERE!
cm = confusion_matrix(y_test, predictions)
print(cm)

[[18655   110]
 [   36   583]]


In [ ]:
# Print the classification report for the model
# YOUR CODE HERE!
print(classification_report(y_test,predictions,target_names=["healthy loan", "high risk loan"]))

                precision    recall  f1-score   support

  healthy loan       1.00      0.99      1.00     18765
high risk loan       0.84      0.94      0.89       619

      accuracy                           0.99     19384
     macro avg       0.92      0.97      0.94     19384
  weighted avg       0.99      0.99      0.99     19384



### Step 4: Answer the following question.

**Question:** How well does the logistic regression model predict both the `0` (healthy loan) and `1` (high-risk loan) labels?

**Answer:** According to the classification report, the model predicts a healthy loan very well based on the 1.00 precision and majority of the dataset was even classified as a healthy loan. As for the high risk loan predictions, the precision and recall scores are still pretty high meaning mis-classification is more likely to occur (compared to healthy loans) but majority are identified correctly. Overall, the model predicts the labels very well!

---

## Predict a Logistic Regression Model with Resampled Training Data

### Step 1: Use the `RandomOverSampler` module from the imbalanced-learn library to resample the data. Be sure to confirm that the labels have an equal number of data points.

In [ ]:
# Import the RandomOverSampler module form imbalanced-learn
from imblearn.over_sampling import RandomOverSampler

# Instantiate the random oversampler model
# # Assign a random_state parameter of 1 to the model
# YOUR CODE HERE!

res = RandomOverSampler(random_state=1)

# Fit the original training data to the random_oversampler model
# YOUR CODE HERE!
X_resampled, y_resampled = res.fit_resample(X_train, y_train)

In [ ]:
# Count the distinct values of the resampled labels data
# YOUR CODE HERE!
counts = pd.Series(y_resampled).value_counts()
counts

loan_status
0    56271
1    56271
Name: count, dtype: int64

### Step 2: Use the `LogisticRegression` classifier and the resampled data to fit the model and make predictions.

In [ ]:
# Instantiate the Logistic Regression model
# Assign a random_state parameter of 1 to the model
# YOUR CODE HERE!
model2 = LogisticRegression(random_state=1)

# Fit the model using the resampled training data
# YOUR CODE HERE!
model.fit(X_resampled,y_resampled)

# Make a prediction using the testing data
# YOUR CODE HERE!
predictions2 = model.predict(X_test)

### Step 3: Evaluate the model’s performance by doing the following:

* Calculate the accuracy score of the model.

* Generate a confusion matrix.

* Print the classification report.

In [ ]:
# Print the balanced_accuracy score of the model
# YOUR CODE HERE!
accuracy_score2 = balanced_accuracy_score(y_test, predictions2)
print(f"Accuracy Score: {accuracy_score2 * 100:.2f}%")


Accuracy Score: 99.36%


In [ ]:
# Generate a confusion matrix for the model
# YOUR CODE HERE!
cm2 = confusion_matrix(y_test, predictions2)
print(cm2)

[[18646   119]
 [    4   615]]


In [ ]:
# Print the classification report for the model
# YOUR CODE HERE!
print(classification_report(y_test,predictions2,target_names=["healthy loan", "high risk loan"]))

                precision    recall  f1-score   support

  healthy loan       1.00      0.99      1.00     18765
high risk loan       0.84      0.99      0.91       619

      accuracy                           0.99     19384
     macro avg       0.92      0.99      0.95     19384
  weighted avg       0.99      0.99      0.99     19384



### Step 4: Answer the following question

**Question:** How well does the logistic regression model, fit with oversampled data, predict both the `0` (healthy loan) and `1` (high-risk loan) labels?

**Answer:** The logistic regression model with oversampled data predicted healthy loans perfectly as shown with a 100% percision and similarily predicted high risk loans well except with a slightly lower percision score (0.84) meaning its only 84% accurate whne it predicts a high-risk loan but still a high recall score meaning its rarely misclassified.